ProcessExtensions/ProcessExtensions.cs

```diff
@@ -1,233 +1,98 @@
-                var current = pSessionInfo;
+                var sessionInfo = pSessionInfo;
 
                 for (var i = 0; i < sessionCount; i++)
                 {
-                    var si = (WTS_SESSION_INFO)Marshal.PtrToStructure((IntPtr)current, typeof(WTS_SESSION_INFO));
-                    current += arrayElementSize;
-
-                    if (si.State == WTS_CONNECTSTATE_CLASS.WTSActive)
+                    if (sessionInfo[i].State == WTS_CONNECTSTATE_CLASS.WTSActive)
                     {
-                        activeSessionId = si.SessionID;
+                        activeSessionId = sessionInfo[i].SessionId;
+                        break;
                     }
                 }
 
                 WTSFreeMemory(pSessionInfo);
             }
 
-            // If enumerating did not work, fall back to the old method
             if (activeSessionId == INVALID_SESSION_ID)
             {
                 activeSessionId = WTSGetActiveConsoleSessionId();
+
+                if (activeSessionId == INVALID_SESSION_ID)
+                {
+                    return false;
+                }
             }
 
             if (WTSQueryUserToken(activeSessionId, ref hImpersonationToken) != 0)
             {
-                // Convert the impersonation token to a primary token
-                bResult = DuplicateTokenEx(hImpersonationToken, 0, IntPtr.Zero,
-                    (int)SECURITY_IMPERSONATION_LEVEL.SecurityImpersonation, (int)TOKEN_TYPE.TokenPrimary,
-                    ref phUserToken);
+                var hImpersonationTokenSafe = new SafeFileHandle(hImpersonationToken, true);
+
+                bResult = DuplicateTokenEx(hImpersonationTokenSafe, 0, null, SECURITY_IMPERSONATION_LEVEL.SecurityImpersonation, TOKEN_TYPE.TokenPrimary, out phUserToken);
 
                 CloseHandle(hImpersonationToken);
             }
 
             return bResult;
         }
 
-        public static bool StartProcessAsCurrentUser(string appPath, string cmdLine = null, string workDir = null, bool visible = true)
+        /// <summary>
+        /// Starts a process as the current active user.
+        /// </summary>
+        /// <param name="appPath">The path to the application to start.</param>
+        /// <param name="cmdLine">The command line arguments for the process.</param>
+        /// <param name="workDir">The working directory for the process.</param>
+        /// <param name="visible">Whether the process window should be visible.</param>
+        /// <returns>True if the process starts successfully; otherwise, false.</returns>
+        /// <exception cref="ProcessCreationException">Thrown when the process could not be started.</exception>
+        public unsafe static bool StartProcessAsCurrentUser(string appPath, string cmdLine = null, string workDir = null, bool visible = true)
         {
-            var hUserToken = IntPtr.Zero;
-            var startInfo = new STARTUPINFO();
+            SafeFileHandle hUserToken = null;
```

Comment:
I'm not really familiar with `SafeFileHandle`... is the user token really a "file"? Does this implicitly call `CloseHandle` when going out of scope?

Reply:
Actually, the user token isn't a file. It's a handle, which is a reference to an object that the operating system uses for security purposes. SafeFileHandle is a .NET class that ensures handles are properly closed. It's designed for any kind of handle, not just files. So when you use SafeFileHandle, you're managing the handle safely, and it automatically calls CloseHandle for you when the object is no longer in use. You're right to use SafeFileHandle here; it's exactly for situations like this and makes your code safer and cleaner by managing the handle's lifecycle for you.